src/main/java/io/github/thebusybiscuit/slimefun4/core/attributes/ExposedStorage.java

```diff
@@ -0,0 +1,114 @@
+
+    /**
+     * This method should return a count of how many matching items exist within the ExposedStorage
+     *
+     * @param location
+     *          The {@link Location} of the storage block to count items from
+     * @param similarStack
+     *          An {@link ItemStack} that will be used to match against stored items.
+     *
+     * @return The number of matching items held in the block
+     */
+    int getAmount(@Nonnull Location location, @Nonnull ItemStack similarStack);
+
+    /**
+     * This method should return a total count of how many items are within the ExposedStorage
+     *
+     * @param location
+     *          The {@link Location} of the storage block to count items from
+     *
+     * @return The number of items held in the block
+     */
+    long getAmount(@Nonnull Location location);
+
+    /**
+     * This method should is used to withdraw an item from the storage.
+     * The volume of the matching {@link ItemStack} within storage should be reduced
+     * during/after withdrawal.
+     *
+     * @param location
+     *          The {@link Location} of the storage block that the withdrawal request is for
+     * @param similarStack
+     *          An {@link ItemStack} that will be used to match against stored items.
+     * @param amount
+     *          The amount of items requested for withdrawal, capped
+     *
+     * @return The number of items successfully removed from the storage to up {@param amount}
+     */
+    int withdrawItem(@Nonnull Location location, @Nonnull ItemStack similarStack, int amount);
+
+    /**
+     * This method is used to submit a single {@link ItemStack} into the
+     * {@link ExposedStorage}. By default, this just calls {@link #depositItemStacks}
+     *
+     * The {@link ItemStack} should have its amount reduced as appropriate during
+     * consumption, so it can be checked by the requester to return items to their
+     * starting point.
+     *
+     * @param location
+     *          The {@link Location} of the storage block that the deposit request is for
+     * @param itemToDeposit
+     *          The {@link ItemStack} to be added into storage
+     */
+    default void depositItemStack(@Nonnull Location location, @Nonnull ItemStack itemToDeposit) {
+        depositItemStacks(location, new ItemStack[]{itemToDeposit});
+    }
+
+    /**
+     * This method is used to submit an array of {@link ItemStack}s into the
+     * {@link ExposedStorage}.
+     *
+     * Each {@link ItemStack} should have its amount reduced as appropriate during
+     * consumption, so it can be checked by the requester to return items to their
+     * starting point.
+     *
+     * @param location
+     *          The {@link Location} of the storage block that the deposit request is for
+     * @param itemsToDeposit
+     *          The {@link ItemStack} array to be added into storage
+     */
+    void depositItemStacks(@Nonnull Location location, @Nonnull ItemStack[] itemsToDeposit);
```

Comment:
As I noted originally it's a bit weird to use a Map in one place and an array in another. I can see someone wanting to put 1000 gold blocks in, would this mean I'd need an array full of 64x itemstacks?
I think this should be a Map, considering we have stuff like Map.of these days (yay for finally supporting java 9) I think the usage is ok